README.md
# Semantic Image Segmentation Unet
**Members** : <a href="https://github.com/vegovs">Vegard Bergsvik Øvstegård</a>

**Supervisors** : <a href="https://www.mn.uio.no/ifi/personer/vit/jimtoer/">Jim Tørresen</a>

## Description

This repository aims to implement and produce trained networks in semantic image segmentation for
[orthopohots](https://www.sciencedirect.com/topics/earth-and-planetary-sciences/orthophoto).
Current network structure is [U-net](https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/).

## Dependencies
* [Python](https://www.python.org/) (version 3.8)
* [Pip](https://virtualenv.pypa.io/en/latest/)
* [conda](https://docs.conda.io/en/latest/)
* [Cuda](https://developer.nvidia.com/cuda-10.2-download-archive) version 10.2

## Installation

```console
git clone https://github.com/gil-uav/semantic-image-segmentation-unet.git
```

#### Virtualenv
```console
python -m virtualevn venv
source venv/bin/activate
pip install -r requirements.txt
```

#### Conda
```console
conda create --name seg --file spec-file.txt
conda activate seg
pip install kornia
```

## Usage

### Training
The application fetches some configurations and parameters from a .env file if it exists.
Run `python train.py --help` to see all other arguments. The package is using [pytorch-lighting](https://github.com/PyTorchLightning/pytorch-lightning) and inherits all its arguments.

The data is expected to be structured like this:
```
data/
    images/
    masks/
```
The path do data us set using --dp argument.

#### Console example
This example stores the checkpoints and logs under the default_root_dir, uses all available GPUs and
fetches training data from --dp.

```console
python train.py --default_root_dir=/shared/use/this/ --gpus=-1 --dp=/data/is/here/
```

#### .env example
Only these arguments are fetched from .env, the rest must be passed through the CLI.
```
# Model config
N_CHANNELS=3
N_CLASSES=1
BILINEAR=True

# Hyperparameters
EPOCHS=300 # Epochs
BATCH_SIZE=4 # Batch size
GROUP_NORM=0 # Group normalization size. If 0 run batch norm.
LRN_RATE=0.001 # Learning rate
VAL_PERC=15 # Validation percent
TEST_PERC=15 # Testing percent
IMG_SIZE=512  # Image size
VAL_INT_PER=1 # Validation interval percentage
ACC_GRAD=4 # Accumulated gradients, number = K.
GRAD_CLIP=1.0 # Clip gradients with norm above give value
EARLY_STOP=10 # Early stopping patience(Epochs)

# Other
PROD=False # Turn on or off debugging APIs
FAST_DEV_RUN=False # Do a fast test.
DIR_DATA="data/" # Where dataset is stored
DIR_ROOT_DIR="/shared/use/this/" # Where logs and checkpoint will be stored
WORKERS=4 # Number of workers for data- and validation loading
DISCORD_WH=httpsomethingwebhoowawnserisalways42
```

### Performance tips:
* Try with different number of workers, but more than 0. A good starting point
is `workers = cores * (threads per core)`.

## Features
### ML:
* Group normalization
* Auto learning rate tuner
* Distributed data parallel training
* Early stopping
* ADAM optimizer
* Gradient clipping
* ReduceLROnPlateau learning rate scheduler
* Logging to Tensorboard
* Metrics:
    * Loss
    * F1
    * Precision
    * Recall
    * Visualise images
* Gradient Accumulation(NB! Might conflict with batch-norm!)

### Ease of use:
* Add hyper-parameters and arguments from console
* Load hyper-parameters from .env
* Training finished notification to Discord

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
[MIT Licence](https://github.com/gil-uav/semantic-image-segmentation/blob/master/LICENSE)

CHANGELOG.md
## 0.2.0 (2021-04-16)

### Feat

- added group normalization and fiex auto tune lr

### Fix

- fix mask datatype and logging issue

## 0.1.0 (2021-04-14)

### Feat

- initial commit


train.py
import logging
import os
import sys
from argparse import ArgumentParser
from datetime import datetime

from knockknock import discord_sender

import torch
from dotenv import load_dotenv
from pytorch_lightning import Trainer
from pytorch_lightning.callbacks import EarlyStopping, LearningRateMonitor
from pytorch_lightning.loggers import TensorBoardLogger
from pytorch_lightning.plugins import DDPPlugin

from torch.backends import cudnn

from unet.unet_model import UNet

load_dotenv()


@discord_sender(webhook_url=os.getenv("DISCORD_WH"))
def main():
    """
    Main training loop.
    """
    parser = ArgumentParser()

    parser = UNet.add_model_specific_args(parser)
    parser = Trainer.add_argparse_args(parser)

    args = parser.parse_args()

    prod = bool(os.getenv("PROD"))
    logging.getLogger(__name__).setLevel(logging.INFO)

    if prod:
        logging.info("Training i production mode, disabling all debugging APIs")
        torch.autograd.set_detect_anomaly(False)
        torch.autograd.profiler.profile(enabled=False)
        torch.autograd.profiler.emit_nvtx(enabled=False)
    else:
        logging.info("Training i development mode, debugging APIs active.")
        torch.autograd.set_detect_anomaly(True)
        torch.autograd.profiler.profile(
            enabled=True, use_cuda=True, record_shapes=True, profile_memory=True
        )
        torch.autograd.profiler.emit_nvtx(enabled=True, record_shapes=True)

    model = UNet(**vars(args))

    logging.info(
        f"Network:\n"
        f"\t{model.hparams.n_channels} input channels\n"
        f"\t{model.hparams.n_classes} output channels (classes)\n"
        f'\t{"Bilinear" if model.hparams.bilinear else "Transposed conv"} upscaling'
    )

    cudnn.benchmark = True  # cudnn Autotuner
    cudnn.enabled = True  # look for optimal algorithms

    early_stop_callback = EarlyStopping(
        monitor="val_loss",
        min_delta=0.00,
        mode="min",
        patience=10 if not os.getenv("EARLY_STOP") else int(os.getenv("EARLY_STOP")),
        verbose=True,
    )

    lr_monitor = LearningRateMonitor()

    run_name = "{}_LR{}_BS{}_IS{}".format(
        datetime.now().strftime("%d-%m-%Y-%H-%M-%S"),
        args.lr,
        args.batch_size,
        args.image_size,
    ).replace(".", "_")

    log_folder = (
        "./logs" if not os.getenv("DIR_ROOT_DIR") else os.getenv("DIR_ROOT_DIR")
    )
    if not os.path.isdir(log_folder):
        os.mkdir(log_folder)
    logger = TensorBoardLogger(log_folder, name=run_name)

    try:
        trainer = Trainer.from_argparse_args(
            args,
            gpus=-1,
            accelerator="ddp",
            plugins=DDPPlugin(find_unused_parameters=False),
            precision=16,
            auto_lr_find="learning_rate"
            if float(os.getenv("LRN_RATE")) == 0.0
            else False,
            logger=logger,
            callbacks=[early_stop_callback, lr_monitor],
            accumulate_grad_batches=1.0
            if not os.getenv("ACC_GRAD")
            else int(os.getenv("ACC_GRAD")),
            gradient_clip_val=0.0
            if not os.getenv("GRAD_CLIP")
            else float(os.getenv("GRAD_CLIP")),
            max_epochs=100 if not os.getenv("EPOCHS") else int(os.getenv("EPOCHS")),
            val_check_interval=0.1
            if not os.getenv("VAL_INT_PER")
            else float(os.getenv("VAL_INT_PER")),
            default_root_dir=os.getcwd()
            if not os.getenv("DIR_ROOT_DIR")
            else os.getenv("DIR_ROOT_DIR"),
            fast_dev_run=True if os.getenv("FAST_DEV_RUN") == "True" else False,
        )
        if float(os.getenv("LRN_RATE")) == 0.0:
            trainer.tune(model)
        trainer.fit(model)
        trainer.test(model)
    except KeyboardInterrupt:
        torch.save(model.state_dict(), "INTERRUPTED.pth")
        logging.info("Saved interrupt")
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)


if __name__ == "__main__":
    main()


unet/unet_modules.py
import torch
import torch.nn as nn
import torch.nn.functional as f


class DoubleConvolution(nn.Module):
    """
    Class used to initialize the conv 3x3, ReLu step.
    """

    def __init__(
        self,
        in_channels: int,
        out_channels: int,
        mid_channels: int = None,
        group_norm: int = 32,
    ):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        mid_channels : int
            Number if mid-layer channels
        group_norm : int
            Number of group to use in group normalization. If 0 use batch norm.
        """
        super().__init__()
        if not mid_channels:
            mid_channels = out_channels
        self.double_conv = nn.Sequential(
            nn.Conv2d(in_channels, mid_channels, kernel_size=3, padding=1, bias=False),
            nn.GroupNorm(group_norm, mid_channels)
            if group_norm > 0
            else nn.BatchNorm2d(mid_channels),
            nn.ReLU(inplace=True),
            nn.Conv2d(mid_channels, out_channels, kernel_size=3, padding=1, bias=False),
            nn.GroupNorm(group_norm, out_channels)
            if group_norm > 0
            else nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),
        )

    def forward(self, x):
        """
        Preforms and returns results from the conv 3x3, ReLu step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.double_conv(x)


class Down(nn.Module):
    """
    Class used to initialize the max pool 2x2 step.
    """

    def __init__(self, in_channels: int, out_channels: int):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        """
        super().__init__()
        self.maxpool_conv = nn.Sequential(
            nn.MaxPool2d(2), DoubleConvolution(in_channels, out_channels)
        )

    def forward(self, x):
        """
        Preforms and returns results from the max pool 2x2 step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.maxpool_conv(x)


class Up(nn.Module):
    """
    Class used to initialize the up-conv 2x2 step.
    """

    def __init__(self, in_channels: int, out_channels: int, bilinear: bool = True):
        super().__init__()
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        bilinear : bool
            Bilinear interpolation in upsampling(default)
        """

        if bilinear:
            self.up = nn.Upsample(scale_factor=2, mode="bilinear", align_corners=True)
            self.conv = DoubleConvolution(in_channels, out_channels, in_channels // 2)
        else:
            self.up = nn.ConvTranspose2d(
                in_channels, in_channels // 2, kernel_size=2, stride=2
            )
            self.conv = DoubleConvolution(in_channels, out_channels)

    def forward(self, x1, x2):
        """
        Preforms and returns results from the up-conv 2x2 step.

        Parameters
        ----------
        x1 : torch.tensor
            From
        x2 : torch.tensor
            To
        """
        x1 = self.up(x1)
        diff_y = x2.size()[2] - x1.size()[2]
        diff_x = x2.size()[3] - x1.size()[3]

        x1 = f.pad(
            x1, [diff_x // 2, diff_x - diff_x // 2, diff_y // 2, diff_y - diff_y // 2]
        )
        x = torch.cat([x2, x1], dim=1)
        return self.conv(x)


class OutConvolution(nn.Module):
    """
    Class used to initialize the conv 1x1 step.
    """

    def __init__(self, in_channels: int, out_channels: int):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        """
        super(OutConvolution, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=1)

    def forward(self, x):
        """
        Preforms and returns results from the conv 1x1 step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.conv(x)


unet/unet_model.py
"""
Pytorch Ignite U-net module.
"""
import copy
import os
import random
from argparse import ArgumentParser

import kornia
import pytorch_lightning as pl
import torch
import torch.nn as nn
import torchvision
from torch.optim.lr_scheduler import ReduceLROnPlateau
from torch.utils.data import DataLoader, random_split

from unet.unet_modules import DoubleConvolution, Down, OutConvolution, Up
from utils.dataset import OrthogonalPhotoDataset


class UNet(pl.LightningModule):
    """Basic U-net structure as described in : O. Ronneberger, P. Fischer, and T. Brox,
    “U-net: Convolutional networks for biomedical image segmentation.” 2015.
    """

    def __init__(
        self,
        n_channels: int = 3,
        n_classes: int = 1,
        bilinear=True,
        learning_rate: float = 0.0001,
        group_norm: int = 0,
        **kwargs,
    ):
        """

        Parameters
        ----------
        n_channels : int
            Number of channels in input-data.
        n_classes : int
            Number of classes to segment.
        bilinear : bool
            Bilinear interpolation in upsampling(default)
        group_norm : int
            Number of group to use in group normalization. If 0 use batch norm.
        """
        super().__init__()
        self.learning_rate = learning_rate
        self.save_hyperparameters()

        self.example_input_array = torch.randn(
            self.hparams.batch_size,
            n_channels,
            self.hparams.image_size,
            self.hparams.image_size,
        )

        if self.hparams.n_classes > 1:
            self.criterion = nn.CrossEntropyLoss()
        else:
            self.criterion = nn.BCEWithLogitsLoss()

        factor = 2 if bilinear else 1

        # Model definition start
        self.in_conv = DoubleConvolution(n_channels, 64, group_norm)
        self.down_conv_1 = Down(64, 128)
        self.down_conv_2 = Down(128, 256)
        self.down_conv_3 = Down(256, 512)
        self.down_conv_4 = Down(512, 1024 // factor)

        self.up_conv_1 = Up(1024, 512 // factor, bilinear)
        self.up_conv_2 = Up(512, 256 // factor, bilinear)
        self.up_conv_3 = Up(256, 128 // factor, bilinear)
        self.up_conv_4 = Up(128, 64, bilinear)
        self.out_conv = OutConvolution(64, n_classes)
        # Model definition end

        # Dataset
        self.train_set = None
        self.val_set = None
        self.test_set = None

        # Metrics
        is_multiclass = True if self.hparams.n_classes > 1 else False
        self.train_pres = pl.metrics.Precision(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.train_rec = pl.metrics.Recall(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.train_f1 = pl.metrics.classification.FBeta(
            num_classes=self.hparams.n_classes
        )

        self.val_pres = pl.metrics.Precision(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.val_rec = pl.metrics.Recall(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.val_f1 = pl.metrics.classification.FBeta(
            num_classes=self.hparams.n_classes
        )

        self.test_pres = pl.metrics.Precision(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.test_rec = pl.metrics.Recall(
            num_classes=self.hparams.n_classes, is_multiclass=is_multiclass
        )
        self.test_f1 = pl.metrics.classification.FBeta(
            num_classes=self.hparams.n_classes
        )

        # Logging iterations
        self.train_log = 0
        self.val_log = 0
        self.test_log = 0

    def forward(self, x):
        """
        Feed-forward function.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        x1 = self.in_conv(x)
        x2 = self.down_conv_1(x1)
        x3 = self.down_conv_2(x2)
        x4 = self.down_conv_3(x3)
        x5 = self.down_conv_4(x4)

        x = self.up_conv_1(x5, x4)
        x = self.up_conv_2(x, x3)
        x = self.up_conv_3(x, x2)
        x = self.up_conv_4(x, x1)
        out = self.out_conv(x)

        return out

    def setup(self, stage):
        """
        Initiates datasets.
        """
        dataset = OrthogonalPhotoDataset(**self.hparams)
        test_dataset = copy.deepcopy(OrthogonalPhotoDataset(**self.hparams))
        test_amount = int(len(dataset) * self.hparams.test_percent / 100)
        val_amount = int(len(dataset) * self.hparams.val_percent / 100)
        train_amount = len(dataset) - val_amount - test_amount

        # Logging iterations
        self.train_log = int(train_amount / 10)
        self.val_log = int(val_amount / 10)
        self.test_log = int(test_amount / 10)

        # Development-mode
        if not os.getenv("PROD"):
            # Fix the generator for reproducible results
            self.train_set, self.val_set, self.test_set = random_split(
                dataset,
                [train_amount, val_amount, test_amount],
                generator=torch.Generator().manual_seed(42),
            )
        # Production-mode
        else:
            # Random generator
            self.train_set, self.val_set, self.test_set = random_split(
                dataset, [train_amount, val_amount, test_amount]
            )

        # Disable augmentation for validation set.
        self.val_set.dataset = test_dataset
        self.test_set.dataset = test_dataset
        self.val_set.dataset.set_augmentation(False)
        self.test_set.dataset.set_augmentation(False)

    def train_dataloader(self):
        """
        Creates dataloader for the training set.
        """
        return DataLoader(
            self.train_set,
            batch_size=self.hparams.batch_size,
            shuffle=True,
            num_workers=int(os.getenv("WORKERS")),
            pin_memory=True,
        )

    def val_dataloader(self):
        """
        Creates dataloader for the validation set.
        """
        return DataLoader(
            self.val_set,
            batch_size=self.hparams.batch_size,
            shuffle=False,
            num_workers=int(os.getenv("WORKERS")),
            pin_memory=True,
            drop_last=True,
        )

    def test_dataloader(self):
        """
        Creates dataloader for the test set.
        """
        return DataLoader(
            self.test_set,
            batch_size=self.hparams.batch_size,
            shuffle=False,
            num_workers=int(os.getenv("WORKERS")),
            pin_memory=True,
            drop_last=True,
        )

    def configure_optimizers(self):
        """
        Configures optimizers and learning rate scheduler.

        Returns
        -------
        dict {
            "optimizer": optimizer,
            "lr_scheduler": lr_scheduler,
            "monitor": "val_loss",
            }

        """
        optimizer = torch.optim.Adam(
            self.parameters(), lr=(self.learning_rate or self.learning_rate)
        )
        lr_scheduler = {
            "scheduler": ReduceLROnPlateau(
                optimizer,
                mode="min",
                factor=0.01,
                patience=10,
                threshold=0.0001,
                threshold_mode="rel",
                cooldown=0,
                min_lr=0,
                eps=1e-08,
                verbose=False,
            ),
            "name": "learning_rate",
            "monitor": "val_loss",
        }

        return {
            "optimizer": optimizer,
            "lr_scheduler": lr_scheduler,
        }

    def loss_function(self, logits, target):
        """
        Loss function.
        """
        return self.criterion(logits, target)

    def log_parameters(self):
        """
        Logs parameters as distribution and histograms.
        """
        for tag, value in self.named_parameters():
            tag = tag.replace(".", "/")
            self.logger.experiment.add_histogram(tag, value, self.current_epoch)

    def log_images(self, images, masks, masks_pred, no_imgs: int, step: str):
        """
        Logs example images.
        """
        no_imgs = max(min(no_imgs, self.hparams.batch_size), 0)
        rand_idx = random.sample(range(0, self.hparams.batch_size), no_imgs)
        msk_list = []
        img_list = []

        alpha = 0.5
        for idx in rand_idx:
            onehot = torch.gt(torch.sigmoid(masks_pred[idx]), 0.5)
            msk_grid = torchvision.utils.make_grid([masks[idx], onehot], nrow=1)
            msk_list.append(msk_grid)
            img_grid = torchvision.utils.make_grid(
                [
                    images[idx],
                    torch.clamp(
                        kornia.enhance.add_weighted(
                            src1=images[idx],
                            alpha=1.0,
                            src2=onehot,
                            beta=alpha,
                            gamma=0.0,
                        ),
                        max=1.0,
                    ),
                ],
                nrow=1,
            )
            img_list.append(img_grid)

        mask_grid = torchvision.utils.make_grid(msk_list, nrow=no_imgs)
        self.logger.experiment.add_image(
            "{} - Target vs Predicted".format(step), mask_grid, self.global_step
        )
        image_grid = torchvision.utils.make_grid(img_list, nrow=no_imgs)
        self.logger.experiment.add_image(
            "{} - Image vs Predicted".format(step), image_grid, self.global_step
        )

    def training_step(self, batch, batch_idx):
        """
        Training step.
        """
        images, masks, masks_pred, loss = self.shared_step(batch)

        if batch_idx % self.train_log == 0:
            self.log_parameters()
            self.log_images(images, masks, masks_pred, 5, "TRAIN")

        pred = torch.gt(torch.sigmoid(masks_pred), 0.5)
        masks = masks.type(torch.cuda.ByteTensor)
        self.train_pres(pred, masks)
        self.train_rec(pred, masks)
        self.train_f1(pred, masks)

        self.log("train_loss", loss)
        self.log("train_precision", self.train_pres)
        self.log("train_recall", self.train_rec)
        self.log("train_f1", self.train_f1)

        return loss

    def validation_step(self, batch, batch_idx):
        """
        Validation step.
        """
        images, masks, masks_pred, loss = self.shared_step(batch)

        if batch_idx % self.val_log == 0:
            self.log_images(images, masks, masks_pred, 5, "VAL")

        pred = torch.gt(torch.sigmoid(masks_pred), 0.5)
        masks = masks.type(torch.cuda.ByteTensor)
        self.val_pres(pred, masks)
        self.val_rec(pred, masks)
        self.val_f1(pred, masks)

        self.log("val_loss", loss)
        self.log("val_precision", self.val_pres)
        self.log("val_recall", self.val_rec)
        self.log("val_f1", self.val_f1)

        return {
            "val_loss": loss,
            "val_precision": self.val_pres,
            "val_recall": self.val_rec,
            "val_f1": self.val_f1,
        }

    def test_step(self, batch, batch_idx):
        """
        Test step.
        """
        images, masks, masks_pred, loss = self.shared_step(batch)

        if batch_idx % self.test_log == 0:
            self.log_images(images, masks, masks_pred, 5, "TEST")

        pred = torch.gt(torch.sigmoid(masks_pred), 0.5)
        masks = masks.type(torch.cuda.ByteTensor)
        self.test_pres(pred, masks)
        self.test_rec(pred, masks)
        self.test_f1(pred, masks)

        self.log("test_loss", loss)
        self.log("test_precision", self.test_pres)
        self.log("test_recall", self.test_rec)
        self.log("test_f1", self.test_f1)

        return {
            "test_loss": loss,
            "test_precision": self.val_pres,
            "test_recall": self.val_rec,
            "test_f1": self.val_f1,
        }

    def shared_step(self, batch):
        """
        Runs forward prop and calculates metrics.
        """
        images, masks = batch["image"], batch["mask"]

        if images.shape[1] != self.hparams.n_channels:
            raise AssertionError(
                f"Network has been defined with {self.n_channels} input channels, "
                f"but loaded images have {images.shape[1]} channels. Please check that "
                "the images are loaded correctly."
            )
        masks_pred = self(images)
        loss = self.loss_function(masks_pred, masks)

        return images, masks, masks_pred, loss

    @staticmethod
    def add_model_specific_args(parent_parser):
        """
        Argument parser.
        """
        parser = ArgumentParser(parents=[parent_parser], add_help=False)
        parser.add_argument(
            "-ch",
            "--n-channels",
            dest="n_channels",
            type=int,
            metavar="NCHS",
            default=3 if not os.getenv("N_CHANNELS") else os.getenv("N_CHANNELS"),
            help="Number of channels in input.",
        )
        parser.add_argument(
            "-cl",
            "--n-classes",
            dest="n_classes",
            type=int,
            metavar="NCLS",
            default=1 if not os.getenv("N_CLASSES") else os.getenv("N_CLASSES"),
            help="Number of classes to classify.",
        )
        parser.add_argument(
            "-gn",
            "--group-normalization",
            dest="group_norm",
            type=int,
            metavar="GN",
            default=16 if not os.getenv("GROUP_NORM") else int(os.getenv("GROUP_NORM")),
            help="Number of groups in group noramlization. If 0 run batch normalization.",
        )
        parser.add_argument(
            "-lr",
            "--learning-rate",
            dest="lr",
            type=float,
            metavar="LR",
            default=0.0001 if not os.getenv("LRN_RATE") else os.getenv("LRN_RATE"),
            help="Learning rate.",
        )
        parser.add_argument(
            "-b",
            "--bilinear",
            dest="bilinear",
            type=bool,
            default=True if not os.getenv("BILINEAR") else os.getenv("BILINEAR"),
            help="To use bilinear up-scaling i in the model.",
        )
        parser.add_argument(
            "-is",
            "--image-size",
            dest="image_size",
            type=int,
            metavar="IS",
            default=512 if not os.getenv("IMG_SIZE") else os.getenv("IMG_SIZE"),
            help="Image height value for resizing.",
        )
        parser.add_argument(
            "-valp",
            "--validation-percent",
            metavar="VALP",
            dest="val_percent",
            type=float,
            default=10.0 if not os.getenv("VAL_PERC") else os.getenv("VAL_PERC"),
            help="How much of dataset to be used as validation set.",
        )
        parser.add_argument(
            "-tesp",
            "--testing-percent",
            metavar="TESP",
            dest="test_percent",
            type=float,
            default=10.0 if not os.getenv("TEST_PERC") else os.getenv("TEST_PERC"),
            help="How much of dataset to be used as testing set.",
        )
        parser.add_argument(
            "-bs",
            "--batch-size",
            dest="batch_size",
            metavar="BS",
            type=int,
            default=1 if not os.getenv("BATCH_SIZE") else os.getenv("BATCH_SIZE"),
            help="Batch size",
        )
        parser.add_argument(
            "-dp",
            "--data-path",
            metavar="DP",
            dest="data_path",
            type=str,
            default="data" if not os.getenv("DIR_DATA") else os.getenv("DIR_DATA"),
            help="Load model from a .pth file",
        )

        return parser
